File: Frontend/android/app/src/main/java/com/funified/bandeja/push/AttendancePushData.java
package com.funified.bandeja.push;

import java.util.Map;

/**
 * PRD 346 — the game reminder that carries the two attendance shade actions.
 *
 * <p>Confirmation is a courtesy signal: the actions post an answer and nothing
 * else. They never join, leave, or reorder anybody.
 */
public final class AttendancePushData {
    private static final String TYPE = "GAME_REMINDER";

    public final String title;
    public final String body;
    public final String gameId;
    public final String confirmActionToken;
    public final String unsureActionToken;
    public final String confirmActionTitle;
    public final String unsureActionTitle;
    public final String confirmedAck;
    public final String unsureAck;

    private AttendancePushData(
        String title,
        String body,
        String gameId,
        String confirmActionToken,
        String unsureActionToken,
        String confirmActionTitle,
        String unsureActionTitle,
        String confirmedAck,
        String unsureAck
    ) {
        this.title = title;
        this.body = body;
        this.gameId = gameId;
        this.confirmActionToken = confirmActionToken;
        this.unsureActionToken = unsureActionToken;
        this.confirmActionTitle = confirmActionTitle;
        this.unsureActionTitle = unsureActionTitle;
        this.confirmedAck = confirmedAck;
        this.unsureAck = unsureAck;
    }

    /**
     * A reminder only offers the actions when the backend signed both tokens.
     * A reminder without them stays an ordinary reminder.
     */
    public static boolean hasAttendanceActions(Map<String, String> data) {
        if (data == null || data.isEmpty()) {
            return false;
        }
        return TYPE.equals(trim(data.get("type")))
            && trim(data.get("gameId")) != null
            && trim(data.get("attendanceActionToken")) != null
            && trim(data.get("attendanceUnsureActionToken")) != null;
    }

    public static AttendancePushData fromMap(Map<String, String> data) {
        if (!hasAttendanceActions(data)) {
            return null;
        }
        return new AttendancePushData(
            trim(data.get("title")),
            trim(data.get("body")),
            trim(data.get("gameId")),
            trim(data.get("attendanceActionToken")),
            trim(data.get("attendanceUnsureActionToken")),
            trim(data.get("confirmActionTitle")),
            trim(data.get("unsureActionTitle")),
            trim(data.get("attendanceConfirmedAck")),
            trim(data.get("attendanceUnsureAck"))
        );
    }

    /**
     * Deliberately the same key {@link DataPushNotificationHelper#notificationId}
     * builds for a reminder, so the 2 h reminder replaces the 24 h one instead of
     * stacking a second card in the shade.
     */
    public int notificationId() {
        return (TYPE + ":" + gameId).hashCode() & 0x7fffffff;
    }

    private static String trim(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }
}

File: Frontend/ios/App/App/AttendancePushData.swift
import Foundation

/// PRD 346 — the reminder that carries the two attendance shade actions.
///
/// Confirmation is a courtesy signal: an answer posts a state and nothing else.
/// Neither action can join, leave or reorder anybody.
struct AttendancePushData {
    static let categoryId = "GAME_REMINDER"
    static let confirmActionId = "confirm"
    static let unsureActionId = "unsure"

    enum Answer {
        case confirmed
        case unsure

        init?(actionIdentifier: String) {
            switch actionIdentifier {
            case AttendancePushData.confirmActionId: self = .confirmed
            case AttendancePushData.unsureActionId: self = .unsure
            default: return nil
            }
        }
    }

    let gameId: String
    let confirmActionToken: String
    let unsureActionToken: String
    let confirmedAck: String?
    let unsureAck: String?

    static func from(userInfo: [AnyHashable: Any]) -> AttendancePushData? {
        guard stringValue(userInfo["type"]) == categoryId else { return nil }

        let data = nestedData(from: userInfo)
        guard let gameId = stringValue(data["gameId"]),
              let confirmToken = stringValue(data["attendanceActionToken"]),
              let unsureToken = stringValue(data["attendanceUnsureActionToken"]) else {
            return nil
        }

        return AttendancePushData(
            gameId: gameId,
            confirmActionToken: confirmToken,
            unsureActionToken: unsureToken,
            confirmedAck: stringValue(data["attendanceConfirmedAck"]),
            unsureAck: stringValue(data["attendanceUnsureAck"])
        )
    }

    func actionToken(for answer: Answer) -> String {
        switch answer {
        case .confirmed: return confirmActionToken
        case .unsure: return unsureActionToken
        }
    }

    /// The backend localizes both acknowledgements per recipient, because the
    /// shade handler answers without a webview to translate in.
    func acknowledgement(for answer: Answer) -> String? {
        switch answer {
        case .confirmed: return confirmedAck
        case .unsure: return unsureAck
        }
    }

    private static func nestedData(from userInfo: [AnyHashable: Any]) -> [String: Any] {
        if let payload = userInfo["data"] as? [String: Any] {
            return payload
        }
        var flat: [String: Any] = [:]
        for (key, value) in userInfo {
            if let key = key as? String {
                flat[key] = value
            }
        }
        return flat
    }

    private static func stringValue(_ value: Any?) -> String? {
        guard let value else { return nil }
        let text = String(describing: value).trimmingCharacters(in: .whitespacesAndNewlines)
        return text.isEmpty ? nil : text
    }
}
